helix-term/src/ui/lsp/hover.rs

```diff
@@ -0,0 +1,203 @@
+            format!(
+                "**[{}/{}] {}**",
+                self.active_index + 1,
+                self.hovers.len(),
+                lsp_name
+            ),
+            self.config_loader.clone(),
+        ));
+        let contents = hover_contents_to_string(&hover.contents);
+        self.contents = Some(Markdown::new(contents, self.config_loader.clone()));
+    }
+
+    pub fn set_hover(&mut self, hovers: Vec<(String, lsp::Hover)>) {
+        self.hovers = hovers;
+        self.set_index(usize::default());
+    }
+
+    fn set_index(&mut self, index: usize) {
+        self.active_index = index;
+        self.prepare_markdowns();
+    }
+
+    pub fn next_hover(&mut self) {
+        let index = if self.active_index < self.hovers.len() - 1 {
+            self.active_index + 1
+        } else {
+            usize::default()
+        };
+        self.set_index(index);
+    }
+
+    pub fn previous_hover(&mut self) {
+        let index = if self.active_index > 0 {
+            self.active_index - 1
+        } else {
+            self.hovers.len() - 1
+        };
+        self.set_index(index);
+    }
+
+    pub fn visible_popup(compositor: &mut Compositor) -> Option<&mut Popup<Self>> {
+        compositor.find_id::<Popup<Self>>(Self::ID)
+    }
+}
+
+const PADDING: u16 = 2;
+const HEADER_HEIGHT: u16 = 1;
+const SEPARATOR_HEIGHT: u16 = 1;
+
+impl Component for Hover {
+    fn render(&mut self, area: Rect, surface: &mut Buffer, cx: &mut Context) {
+        let margin = Margin::horizontal(1);
+        let area = area.inner(&margin);
+
+        let (Some(header), Some(contents)) = (self.header.as_ref(), self.contents.as_ref()) else {
+            log::info!("markdown not ready");
+            return;
+        };
+
+        // header LSP Name
+        let header = header.parse(Some(&cx.editor.theme));
+        let header = Paragraph::new(&header);
+        header.render(area.with_height(HEADER_HEIGHT), surface);
+
+        // border
+        let sep_style = Style::default();
+        let borders = BorderType::line_symbols(BorderType::Plain);
+        for x in area.left()..area.right() {
+            if let Some(cell) = surface.get_mut(x, area.top() + HEADER_HEIGHT) {
+                cell.set_symbol(borders.horizontal).set_style(sep_style);
+            }
+        }
+
+        // hover content
+        let contents = contents.parse(Some(&cx.editor.theme));
+        let contents_area = area
+            .clip_top(2)
+            .clip_bottom(u16::from(cx.editor.popup_border()));
+        let contents_para = Paragraph::new(&contents)
+            .wrap(Wrap { trim: false })
+            .scroll((cx.scroll.unwrap_or_default() as u16, 0));
+        contents_para.render(contents_area, surface);
+    }
+
+    fn required_size(&mut self, viewport: (u16, u16)) -> Option<(u16, u16)> {
+        if PADDING >= viewport.1 || PADDING >= viewport.0 {
+            return None;
+        }
+        let max_text_width = (viewport.0 - PADDING).min(120);
```

Comment:
Looks like https://github.com/helix-editor/helix/commit/93e8c16614b26278a1a979a3a01a46c9ba3d681f should be applied here since this impl was based on `SignatureHelp`
```suggestion
        let max_text_width = viewport.0.saturating_sub(PADDING).clamp(10, 120);
```